src/guiguts/widgets.py

```diff
@@ -678,6 +679,174 @@ def destroy(self) -> None:
             self.tooltip_window.destroy()
 
 
+@dataclass
+class CommandInfoMetadata:
+    """Class to store info about a command."""
+
+    label: str
+    menu: str
+    shortcut: str
+    command: Callable
+
+    def __lt__(self, other: "CommandInfoMetadata") -> bool:
+        """Define "<" to support sorting by label."""
+        return self.label < other.label
+
+
+class EntryMetadata:
+    """Store metadata about a menu entry."""
+
+    def __init__(self, label: str) -> None:
+        """Initialize EntryMetadata object."""
+        self.label = label
+
+
+class CommandMetadata(EntryMetadata):
+    """Store metadata about a menu button (command)."""
+
+    def __init__(self, label: str, command: Callable, shortcut: str) -> None:
+        """Initialize ButtonMetadata object.
+
+        Args:
+            label: The text displayed for the button.
+            command: The function to execute when clicked.
+            shortcut: Keyboard shortcut for the command (empty string if none)
+        """
+        super().__init__(label)
+        self.command = command
+        self.shortcut = shortcut
+
+
+class MenuMetadata(EntryMetadata):
+    """Store metadata about a menu item."""
+
+    def __init__(self, label: str, widget: tk.Menu) -> None:
+        """Initialize MenuMetadata object.
+
+        Args:
+            label: Label string for menu button.
+            widget: The menu that is being added.
+        """
+        super().__init__(label)
+        self.entries: list[EntryMetadata] = []
+        self.widget = widget
+
+    def get_submenu_metadata(self, submenu: tk.Menu) -> Optional["MenuMetadata"]:
+        """Return the metadata for the given submenu."""
+        for entry in self.entries:
+            if isinstance(entry, MenuMetadata):
+                if entry.widget == submenu:
+                    return entry  # Found it!
+                # Also recurse into submenus
+                if metadata := entry.get_submenu_metadata(submenu):
+                    return metadata
+        return None
+
+    def add_button(self, label: str, command: Callable, shortcut: str = "") -> None:
+        """Add a command button to this menu."""
+        button = CommandMetadata(label, command, shortcut)
+        self.entries.append(button)
+
+    def get_all_commands(self) -> list[CommandInfoMetadata]:
```

Comment:
Using vscode as an example, I often want to run the same command I just ran. Perhaps we can store the last-used command somewhere and place it at the top of the stack instead of in its sorted place?

Even better would be to keep the most recent handful (5?) but that would require more storage locations, a rotation and cleanup algorithm, etc... I'd consider just the most recent one a win.

Reply:
I'll look into that